url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/sling/SlingParser.java

```diff
@@ -0,0 +1,522 @@
+ * <br>
+ * baseElement = 
+ *     Num | "pi" | "random" | "s1" | "s2" | "t" | variable | 
+ *     oneArg("abs")    | oneArg("ceil")       | 
+ *     oneArg("cos")    | oneArg("floor")      | 
+ *     oneArg("sin")    | oneArg("tan")        |
+ *     twoArgs("polar") | twoArgs("cartesian") | 
+ *     twoArgs("scale") | twoArgs("sling");
+ * <br>
+ * oneArg(i)  = i '(' expression ')';
+ * twoArgs(i) = i '(' expression ',' expression ')';
+ * 
+ * </pre></blockquote>
+ * 
+ * The following program describes about 10,000 interesting
+ * plots:
+ *
+ * <blockquote><pre>
+ *     plot sling(1, 1) + sling(s1, 100*s2);
+ * </pre></blockquote>
+ *
+ * <p>
+ * The class <code>SlingIde</code> provides an interactive
+ * development environment for Sling.
+ *
+ * @author Steven J. Metsker
+ *
+ * @version 1.0 
+ */
+public class SlingParser {
+	protected Sequence expression;
+	protected Alternation statement;
+	protected Alternation baseElement;
+	protected WordOrReservedState wors;
+	protected Tokenizer tokenizer;
+
+	/*
+	 * Returns a parser that will recognize the grammar:
+	 * 
+	 *     assignment = variable '=' expression ';' ; 
+	 */
+	protected Parser assignment() {
+		Track t = new Track("assignment");
+		t.add(variable());
+		t.add(new Symbol('=').discard());
+		t.add(expression());
+		t.add(new Symbol(';').discard());
+		t.setAssembler(new AssignmentAssembler());
+		return t;
+	}
+
+	/*
+	 * Returns a parser that will recognize the grammar:
+	 * 
+	 *     baseElement = 
+	 *        Num | "pi" | "random" | "s1" | "s2" | "t" | variable |
+	 *       ("abs" |"ceil" |"cos" | "floor" |"sin" |"tan") oneArg |
+	 *       ("polar" | "cartesian" | "scale" | "sling") twoArgs;
+	 */
+	protected Parser baseElement() {
+		if (baseElement == null) {
+			baseElement = new Alternation("base elements");
+			baseElement.add(oneArg("abs", new Abs()));
+			baseElement.add(twoArg("cartesian", new Cartesian()));
+			baseElement.add(oneArg("ceil", new Ceil()));
+			baseElement.add(oneArg("cos", new Cos()));
+			baseElement.add(oneArg("floor", new Floor()));
+			baseElement.add(num());
+			baseElement.add(noArgs("random", new Random()));
+			baseElement.add(pi());
+			baseElement.add(twoArg("polar", new Polar()));
+			baseElement.add(s1());
+			baseElement.add(s2());
+			baseElement.add(scale());
+			baseElement.add(oneArg("sin", new Sin()));
+			baseElement.add(twoArg("sling", new Sling()));
+			baseElement.add(noArgs("t", new T()));
+			baseElement.add(oneArg("tan", new Tan()));
+			baseElement.add(variable());
+		}
+		return baseElement;
+	}
+
+	/*
+	 * Recognize a comma.
+	 */
+	protected static Parser comma() {
+		return new Symbol(',').discard();
+	}
+
+	/*
+	 * Returns a parser that will recognize the grammar:
+	 * 
+	 *     divideElement  = '/' element;
+	 */
+	protected Track divideElement() {
+		Track t = new Track();
+		t.add(new Symbol('/').discard());
+		t.add(element());
+		t.setAssembler(new FunctionAssembler(new Arithmetic('/')));
+		return t;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 135:139, 231:235, 315:319, 335:339, 466:470.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.